[ENH] Move intersection functions for DatetimeIndex and TimedeltaIndex to Datetimelike and added tests

fdroessler opened the pull request:
- [x] closes #24966
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Because DatetimeIndex and TimedeltaIndex intersections are similar, I have moved the intersection function of both classes up into Datetimelike after discussions with @reidy-p @jorisvandenbossche. At the same time for PeriodIndex I have used Index.intersection instead. Additionally I have added some tests to TimedeltaIndex recycling some tests from DatetimeIndex.

fdroessler committed:
Moving intersection method from DatetimeIndex and Timedeltaindex to datetimelike. Point PeriodIndex intersection to Index.intersection

fdroessler committed:
add tests for timedeltaindex

fdroessler committed:
add whatsnew

fdroessler committed:
fixing isort issue

fdroessler committed:
isort issue

jreback commented on pandas/core/indexes/datetimelike.py:
```diff
@@ -530,6 +531,83 @@ def isin(self, values):
         return algorithms.isin(self.asi8, values.asi8)
 
+    def intersection(self, other, sort=False):
+        """
```
instead of a combined doc-string here, put it in the sub-classes (and call super). @jbrockmendel might have a better way (shared docs) of how we do this.

jorisvandenbossche replied:
Is that worth it just to avoid the "DatetimeIndex or TimedeltaIndex" ?

jreback replied:
yes, we don't have combined doc-strings elsewhere so why start here

jreback commented on pandas/core/indexes/datetimelike.py:
```diff
@@ -530,6 +531,83 @@ def isin(self, values):
+            return self
+        if len(other) == 0:
+            return other
+        # to make our life easier, "sort" the two ranges
```
blank line

jreback commented on pandas/core/indexes/datetimelike.py:
```diff
@@ -530,6 +531,83 @@ def isin(self, values):
+              not other.freq.isAnchored() or
+              (not self.is_monotonic or not other.is_monotonic)):
+            result = Index.intersection(self, other, sort=sort)
+            # Invalidate the freq of `result`, which may not be correct at
```
blank line

jreback commented on pandas/core/indexes/period.py:
```diff
@@ -802,6 +802,9 @@ def join(self, other, how='left', level=None, return_indexers=False,
             return self._apply_meta(result), lidx, ridx
         return self._apply_meta(result)
 
+    def intersection(self, other, sort=False):
```
you might be able to do with a Substitution doc-string on a shared doc

jorisvandenbossche replied:
I think you can do a `@Appender(Index.intersection.__doc__)`

jreback commented on pandas/tests/indexes/timedeltas/test_setops.py:
```diff
@@ -73,3 +74,80 @@ def test_intersection_bug_1708(self):
+            assert result.freq == expected.freq
+
+    @pytest.mark.parametrize("sort", [None, False])
+    def intersection_non_monotonic(self, sort):
```
can you parametrize this

jreback commented on pandas/tests/indexes/timedeltas/test_setops.py:
```diff
@@ -73,3 +74,80 @@ def test_intersection_bug_1708(self):
+        tm.assert_index_equal(index_2, inter_2)
+
+    @pytest.mark.parametrize("sort", [None, False])
+    def test_intersection(self, sort):
```
also parameterize this

fdroessler replied:
Something like
```
@pytest.mark.parametrize("rng, exp", [(timedelta_range('1 day', periods=5, freq='h', name='idx'), timedelta_range('1 day', periods=4, freq='h', name='idx')) ...
```

?

jreback replied:
yep exactly (exp -> expected)

jreback commented on pandas/tests/indexes/timedeltas/test_setops.py:
```diff
@@ -73,3 +74,80 @@ def test_intersection_bug_1708(self):
+        inter = first.intersection(first, sort=sort)
+        assert inter is first
+
+    @pytest.mark.parametrize("sort", [None, False])
```
parametrize this

jorisvandenbossche replied:
What do you mean exactly? (as you are commenting this on a 'parametrize' line)

fdroessler replied:
Is that and the same comments below meant as a way to parameterize the index ranges in the tests? I can understand it in the tests below but not sure what you mean in this case.

jreback replied:
yes same comment as above, as the cases in the parameterization (in addition to this one)

jorisvandenbossche commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -20,7 +20,7 @@ Other Enhancements
 
 - :meth:`Timestamp.replace` now supports the ``fold`` argument to disambiguate DST transition times (:issue:`25017`)
--
+- :meth:`DatetimeIndex.intersection` and :meth:`TimedeltaIndex.intersection` have been moved to :meth:`Datetimelike.intersection` and tests were added for :meth:`TimedeltaIndex.intersection` (:issue:`24966`).
```
We indeed don't need to mention this move to datetimelike. But, by doing that, we are actually solving bugs / enhancements, so it are those that we should mention. 
I think we can mention that TimedeltaIndex.intersection also got a `sort` keyword in the "other enhancements" section. And in the bug fixes section mention some of the bugs we are fixing with this change?

jreback replied:
right that is fine

jorisvandenbossche commented on pandas/core/indexes/datetimelike.py:
```diff
@@ -530,6 +531,83 @@ def isin(self, values):
+
+               Changed the default to ``False`` to match the behaviour
+               from before 0.24.0.
+
```
Can you add a 

```
.. versionchanged:: 0.25.0
    
    The `sort` keyword has been added to TimedeltaIndex as well.
```

fdroessler committed:
update whatsnew

fdroessler committed:
update docstrings

fdroessler committed:
parametrize tests

fdroessler commented:
@jreback and @jorisvandenbossche Thanks for the feedback, I think that all should be addressed, locally it looks find on my side. I hope the `super` syntax is correct and please let me know if the `parametrize` line breaks are offending anyones aesthetic feelings :)

fdroessler committed:
add compatibility of super with python2

fdroessler committed:
isort test_setops

jreback commented:
would love to have this. pls ping if you are interested in continuing. would need to rebase on master.

reidy-p commented:
@fdroessler let me know if you don't have time to finish this at the moment and I can pick it up.

fdroessler commented:
@reidy-p No I am happy to finish it, as mentioned in the comments above I thought I addressed all points from @jreback but maybe I did not mark them appropriately or he can let me know what is missing? I'll rebase and push it.

jorisvandenbossche commented:
@fdroessler can you open a new PR ? (once you push to a branch after the PR has been closed, github does not let you reopen the PR ...)

fdroessler commented:
@jorisvandenbossche @jreback I submitted a new PR #25913 let me know if there are any missing issues you want me to resolve, happy to work on it.
